File: array/python/SetMatrixZeroes.py
# Time: O((m * n) + (m * n)) ~ O(m * n)
# Space: O(1)

def set_zeroes(matrix):
    m, n = len(matrix), len(matrix[0])
    col = False
    for i in range(m):
        if matrix[i][0] == 0:    col = True
        for j in range(n):
            if matrix[i][j] == 0:    matrix[i][0] = matrix[0][j] = 0
    for i in range(m - 1, -1, -1):
        for j in range(n - 1, 0, -1):
            if not matrix[i][0] or not matrix[0][j]:
                matrix[i][j] = 0
        if col:    matrix[i][0] = 0

if __name__=='__main__':
    matrix = [[1],[0]]
    set_zeroes(matrix)
    for m in matrix:    print(m)

File: array/cpp/SetMatrixZeroes.cpp
#include <iostream>
#include <vector>
using namespace std;

void setZero(vector<vector<int>>& matrix){
    int m = matrix.size();
    int n = matrix[0].size();
    bool col = false;
    for (int i = 0; i<m; i++){
        if (matrix[i][0] == 0)
            col = true;
        for (int j = 0; j<n; j++){
            if (matrix[i][j] == 0){
                matrix[i][0] = 0;
                matrix[0][j] = 0;
            }
        }
    }
    
    for (int i = m-1; i>-1; i--){
        for (int j = n-1; j>0; j--){
            if (matrix[i][0] != 0 || matrix[0][j] != 0){
                matrix[i][j] = 0;
            }
            if (col)
                matrix[i][0] = 0;
        }
    }
}

int main(){
    vector<vector<int>> matrix {{1},{0}};
    setZero(matrix);
    for (int i =0; i<matrix.size(); i++){
        for (int ele : matrix[i]){
            cout<< ele<<" ";
        }
        cout<<endl;
    }
    return 0;
}